### PR TITLE
[FEATURE] [WIP] PolarsExecutionEngine

### PR DESCRIPTION
**Pull request by austiezr**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- Creates a `PolarsExecutionEngine` at functional parity to the `PandasExecutionEngine`
- Creates splitter & sampler for `PolarsExecutionEngine`
- Implements `build_polars_validator_with_data` and updates various `self_check/util.py` methods to support testing and functionality at parity with `PandasExecutionEngine`
- Updates core metric providers to support `polars`
- Implements a handful of `_polars` metrics to enable expectations / metrics for the same core tests used in testing `PandasExecutionEngine`
- Adds `--polars` flag for pytest
- Introduces tests for `PolarsExecutionEngine` at parity to the `PandasExecutionEngine`
- Introduces tests for splitting & sampling with `polars` at parity to `pandas`
- Updates `expectation.py` and `expectation_diagnostics` to enable testing and diagnostics of `polars` in custom expectations
- Introduces a `_polars` metric for a contrib expectation to support contributor docs
- Restricts testing expectations with `polars` in V2
- Enables `only_for` and `suppress_test_for` for `polars in V3
- Updates the test definition for `expect_column_values_to_be_unique` as an example of the tests supported by `polars`
- Updates `azure-pipelines-dev.yml` to support testing of Polars in CI/CD
- Creates `reqs/requirements-dev-polars.txt` and includes the same in `requirements-dev.txt`

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
- #5503 

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

**Comment by austiezr** [comment]
Note: All tests pass locally but are currently blocked in pipeline due to the `polars` dependency not being fulfilled.

File: great_expectations/execution_engine/polars_execution_engine.py [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+            return {"reader_method": "read_pickle"}
+        elif path.endswith(".feather") or path.endswith(".ipc"):
+            return {"reader_method": "read_ipc"}
+        elif path.endswith(".avro"):
+            return {"reader_method": "read_avro"}
+
+        else:
+            raise ge_exceptions.ExecutionEngineError(
+                f'Unable to determine reader method from path: "{path}".'
+            )
```
Comment (anthonyburdi) [inline great_expectations/execution_engine/polars_execution_engine.py:430]: I think Abe's note here is correct (for an execution, not a v2 Datasource which references this datasource method in the CLI), and in addition to making it private we can probably move this method up to the parent ExecutionEngine, & add type hints.

Comment (austiezr) [reply great_expectations/execution_engine/polars_execution_engine.py:430]: Agreed; would it make sense to roll that refactor out for all execution engines as a part of this work, or should we extract that into a separate piece of work?

Comment (austiezr) [reply great_expectations/execution_engine/polars_execution_engine.py:430]: Just surfacing this Q again 🙇

**Comment by austiezr** [comment]
Haven't looked to see what lift would be like, but it occurred to me that this work was all pre-PK/ID

File: great_expectations/execution_engine/polars_execution_engine.py [hunk 2]
```diff
@@ -0,0 +1,640 @@
...
+            reader_fn = getattr(pl, reader_method)
+            if reader_options:
+                reader_fn = partial(reader_fn, **reader_options)
+            return reader_fn
+        except AttributeError:
+            raise ge_exceptions.ExecutionEngineError(
+                f'Unable to find reader_method "{reader_method}" in polars.'
+            )
+
+    def resolve_metric_bundle(
```
Comment (anthonyburdi) [inline great_expectations/execution_engine/polars_execution_engine.py:467]: Polars also supports [lazy evaluation](https://pola-rs.github.io/polars-book/user-guide/quickstart/intro.html#lazy-quick-start), maybe we can leverage that in a similar way to our spark execution engine ([bundling](https://docs.greatexpectations.io/docs/terms/execution_engine/#deferred-metrics))? Might be out of the scope of this effort.

**Comment by Bidek56** [comment]
@austiezr any reason this PR was closed? 
Polars is very popular these days and this would be a great addition. Thx

**Comment by Durga854** [comment]
Is this completed or still in development? Why was it closed?

**Comment by austiezr** [comment]
Hey @Bidek56 / @Durga854 -- GX determined there wasn't enough demand for this engine at the time to justify the lift on finishing & integrating. As the underlying arch for engines has shifted significantly since this time, I doubt this would plug-and-play anymore, and likely needs a significant rework. No longer with GX, so can't offer any insight on the potential of this coming in the future!

**Pull request closed without merging**